### PR TITLE
Reactive syncing metrics

### PR DESCRIPTION
**Pull request by nazar-pc**
This PR untangles syncing metrics and makes them reactive, the way metrics are supposed to be in general.

Syncing metrics were bundled in a way that caused coupling across multiple layers: justifications metrics were defined and managed by `ChainSync`, but only updated periodically on tick in `SyncingEngine`, while actual values were queried from `ExtraRequests`. This convoluted architecture was hard to follow when I was looking into https://github.com/paritytech/polkadot-sdk/issues/5333.

Now metrics that correspond to each component are owned by that component and updated as changes are made instead of on tick every 1100ms.

This does add some annoying boilerplate that is a bit harder to maintain, but it separates metrics more nicely and if someone queries them more frequently will give arbitrary resolution. Since metrics updates are just atomic operations I do not expect any performance impact of these changes.

Will add prdoc if changes look good otherwise.

P.S. I noticed that importing requests (and corresponding metrics) were not cleared ever since corresponding code was introduced in https://github.com/paritytech/polkadot-sdk/commit/dc41558b6ef69aad59bb4cfcd9c19c91c1d5c3a9#r145518721 and I left it as is to not change the behavior, but it might be something worth fixing.

cc @dmitry-markin

### CODE REVIEW

File: substrate/client/network/sync/src/justification_requests.rs [hunk 1]
```diff
@@ -146,6 +211,10 @@ impl<B: BlockT> ExtraRequests<B> {
 			}
 			self.failed_requests.entry(request).or_default().push((who, Instant::now()));
 			self.pending_requests.push_front(request);
+			if let Some(metrics) = &self.metrics {
+				metrics.failed.set(self.failed_requests.len().try_into().unwrap_or(u64::MAX));
+				metrics.pending.inc();
+			}
 		} else {
 			trace!(target: LOG_TARGET,
```
Comment (dmitry-markin) [inline substrate/client/network/sync/src/justification_requests.rs:219]: This is out of scope of this PR, but it looks like we can legitimately hit this if the response was delivered after the peer disconnected event arrived. This can happen depending on the order different protocols/streams are polled as these events are emitted by different objects (notifications protocol for sync peers and request-response protocol for responses).

Created an issue fr this: https://github.com/paritytech/polkadot-sdk/issues/5414.

File: substrate/client/network/sync/src/justification_requests.rs [hunk 2]
```diff
@@ -117,6 +170,10 @@ impl<B: BlockT> ExtraRequests<B> {
 	pub(crate) fn peer_disconnected(&mut self, who: &PeerId) {
 		if let Some(request) = self.active_requests.remove(who) {
 			self.pending_requests.push_front(request);
+			if let Some(metrics) = &self.metrics {
+				metrics.active.dec();
+				metrics.pending.inc();
```
Comment (dmitry-markin) [inline substrate/client/network/sync/src/justification_requests.rs:175]: Possible race condition (see the comment in `on_response()`), but the bug is already present without this PR.

File: substrate/client/network/sync/src/justification_requests.rs [hunk 3]
```diff
@@ -301,8 +376,17 @@ impl<'a, B: BlockT> Matcher<'a, B> {
 		for requests in self.extras.failed_requests.values_mut() {
 			requests.retain(|(_, instant)| instant.elapsed() < EXTRA_RETRY_WAIT);
 		}
+		if let Some(metrics) = &self.extras.metrics {
+			metrics
+				.active
+				.set(self.extras.failed_requests.len().try_into().unwrap_or(u64::MAX));
```
Comment (dmitry-markin) [inline substrate/client/network/sync/src/justification_requests.rs:382]: Seems strange to set active requests count to failed requests count.

**Review by dmitry-markin — APPROVED** [review]
Looks good, thank you! With updating metrics on every operation it's more difficult to check every update is handled. Hope you have searched through the code and made sure everything is covered.

File: substrate/client/network/sync/src/justification_requests.rs (hunk 2 above)
Comment (nazar-pc) [reply substrate/client/network/sync/src/justification_requests.rs:175]: Yep, but it is narrow and only affects metrics, so :man_shrugging: 

I was thinking about changing these metrics to counters instead (so you not only know current state, but also number of transitions between metrics polling), but that would break metrics "interface", so I didn't change anything that affects the logic.

**Review by nazar-pc — COMMENTED** [review]
> Hope you have searched through the code and made sure everything is covered.

I checked all occurrences and invariants and I don't think I missed anything.

File: substrate/client/network/sync/src/justification_requests.rs [hunk 4]
```diff
@@ -130,13 +187,21 @@ impl<B: BlockT> ExtraRequests<B> {
 		// currently enforced by the outer network protocol before passing on
 		// messages to chain sync.
 		if let Some(request) = self.active_requests.remove(&who) {
+			if let Some(metrics) = &self.metrics {
+				metrics.active.dec();
+			}
```
Comment (lexnv) [inline substrate/client/network/sync/src/justification_requests.rs:192]: Would grouping the request hashmaps (or vec) with metrics help here?
Some time in the future we might do a `active_requests.remove()` and forget to decrement the appropriate metric.

One suggestion might be to group them into a wrapper:
```rust
struct ActiveRequestsMetered {
    inner: HashMap<..>, // Current self.active_requests
    metrics: Option<Metrics>,
}
```

Maybe an even simpler approach would be to introduce a `fn active_requests_remove() { ...; metrics.active.dec(); }`.
Could also be a followup if would take too long to implement :D

**Review by lexnv — APPROVED** [review]
LGTM! Thanks for contributing!

File: substrate/client/network/sync/src/justification_requests.rs (hunk 4 above)
Comment (nazar-pc) [reply substrate/client/network/sync/src/justification_requests.rs:192]: There are several of these in the code and I'm not sure how much it would help with ergonomics to be completely honest. Metrics are one of those things that are just inherently invasive unfortunately.

**Comment by nazar-pc** [comment]
I resolved minor merge conflict, but also had to restore useless tick. Without this tick tests like `sync::multiple_requests_are_accepted_as_long_as_they_are_not_fulfilled` never finish. I didn't figure out yet what is not being polled correctly without the tick, but something just hangs in syncing engine and is never able to make it past this:
https://github.com/paritytech/polkadot-sdk/blob/9fecd89880bebec86815e3413327dd8a5f80befd/substrate/client/network/test/src/sync.rs#L1020-L1024

If someone with more knowledge in this area can take a look at it that'd be great, but it shouldn't block this PR anymore.
